rust/arrow/src/compute/kernels/concat.rs

```diff
@@ -0,0 +1,395 @@
+            concat_raw_primitive::<TimestampNanosecondType>(array_data_list)
+        }
+        DataType::Interval(IntervalUnit::YearMonth) => {
+            concat_raw_primitive::<IntervalYearMonthType>(array_data_list)
+        }
+        DataType::Interval(IntervalUnit::DayTime) => {
+            concat_raw_primitive::<IntervalDayTimeType>(array_data_list)
+        }
+        DataType::Duration(TimeUnit::Second) => {
+            concat_raw_primitive::<DurationSecondType>(array_data_list)
+        }
+        DataType::Duration(TimeUnit::Millisecond) => {
+            concat_raw_primitive::<DurationMillisecondType>(array_data_list)
+        }
+        DataType::Duration(TimeUnit::Microsecond) => {
+            concat_raw_primitive::<DurationMicrosecondType>(array_data_list)
+        }
+        DataType::Duration(TimeUnit::Nanosecond) => {
+            concat_raw_primitive::<DurationNanosecondType>(array_data_list)
+        }
+        t => unimplemented!("Concat not supported for data type {:?}", t),
+    }
+}
+
+fn concat_primitive<T>(array_data_list: &Vec<ArrayDataRef>) -> Result<ArrayRef>
+where
+    T: ArrowPrimitiveType,
+{
+    let rows_count = array_data_list.iter().map(|a| a.len()).sum::<usize>();
+    let mut builder = PrimitiveBuilder::<T>::new(rows_count);
+
+    for array_data in array_data_list {
+        let a = PrimitiveArray::<T>::from(array_data.clone());
+        for i in 0..a.len() {
+            if a.is_valid(i) {
+                builder.append_value(a.value(i))?;
+            } else {
+                builder.append_null()?;
+            }
+        }
+    }
+
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+// for better performance, we manually concat primitive value buffers instead of using
+// PrimitiveBuilder
+fn concat_raw_primitive<T>(array_data_list: &Vec<ArrayDataRef>) -> Result<ArrayRef>
+where
+    T: ArrowPrimitiveType,
+{
+    let value_count = array_data_list.iter().map(|a| a.len()).sum::<usize>();
+    let mut null_count = 0;
+    let mut values_builder = BufferBuilder::<T>::new(value_count);
+    let mut null_bit_builder = BooleanBufferBuilder::new(value_count);
+
+    for array_data in array_data_list {
+        null_count += array_data.null_count();
+        let value_buffer = &array_data.buffers()[0];
+        values_builder.write_bytes(value_buffer.data(), value_buffer.len())?;
+        for i in 0..array_data.len() {
+            null_bit_builder.append(array_data.is_valid(i))?;
+        }
+    }
+
+    Ok(Arc::new(PrimitiveArray::<T>::from(
+        ArrayData::builder(T::get_data_type())
+            .len(value_count)
+            .add_buffer(values_builder.finish())
+            .null_count(null_count)
+            .null_bit_buffer(null_bit_builder.finish())
+            .build(),
+    )))
+}
+
+// for better performance, we manually concat string value buffers instead of using StringBuilder
+fn concat_raw_string(array_data_list: &Vec<ArrayDataRef>) -> Result<ArrayRef> {
```

Comment:
Would this function handle array slices correctly? May you please test this, either in a new test case or by slicing the second string array in `test_concat_string_arrays`?

Reply:
By handling array slices did you mean StringArray created from str array slice?

Reply:
Sorry, I meant slicing the second string array, something like below, which slices out "1" and "6" from the second array:

```rust
    fn test_concat_string_arrays() -> Result<()> {
        let arr = concat(&vec![
            Arc::new(
                StringArray::try_from(vec![Some("hello"), Some("world")])
                    .expect("Unable to create string array"),
            ) as ArrayRef,
            StringArray::from(vec!["1", "2", "3", "4", "6"]).slice(1, 3),
            Arc::new(
                StringArray::try_from(vec![Some("foo"), Some("bar"), None, Some("baz")])
                    .expect("Unable to create string array"),
            ) as ArrayRef,
        ])?;

        let expected_output = Arc::new(
            StringArray::try_from(vec![
                Some("hello"),
                Some("world"),
                // Some("1"),
                Some("2"),
                Some("3"),
                Some("4"),
                // Some("6"),
                Some("foo"),
                Some("bar"),
                None,
                Some("baz"),
            ])
            .expect("Unable to create string array"),
        ) as ArrayRef;

        assert!(
            arr.equals(&(*expected_output)),
            "expect {:#?} to be: {:#?}",
            arr,
            &expected_output
        );

        Ok(())
    }
```

The same might need to be verified with numeric arrays because we write the whole data in `write_slice`, so if there's an offset, we might get unexpected outputs.

I'll check your branch out locally and check this.

Reply:
I've checked, we get this failure below:

The same test failure occurs if you slice a numeric array.

```
thread 'compute::kernels::concat::tests::test_concat_string_arrays' panicked at 'expect StringArray
[
  "hello",
  "world",
  "1",
  "2",
  "3",
  "46f",
  "oob",
  null,
  "arb",
] to be: StringArray
[
  "hello",
  "world",
  "2",
  "3",
  "4",
  "foo",
  "bar",
  null,
  "baz",
]', arrow/src/compute/kernels/concat.rs:296:9
```

Reply:
I'm nearly done with `append_data()`, which handles offsets. Is it worthwhile to wait for it? I can put up a draft PR tonight

Reply:
@houqp here's a WIP of the changes: https://github.com/apache/arrow/compare/master...nevi-me:ARROW-9007

I'm still testing the string and nested array types, but the primitive types behave fine with offsets

Reply:
Yeah, let's wait for that. I was going to refactor this code by implementing append_data. Good to avoid duplicate efforts :D

Reply:
@houqp I need concat support, so I'm going to temporarily merge your code into a fork of mine. Hopefully we can merge this, #7306  and #7365 in the coming days :)

Reply:
great work on #7365, looking forward to refactor/simplify this PR after it's merged!